File: apps/demo-be-java-vertx/src/main/java/com/organiclever/demojavx/repository/UserRepository.java
package com.organiclever.demojavx.repository;

import com.organiclever.demojavx.domain.model.User;
import io.vertx.core.Future;
import java.util.List;
import java.util.Optional;
import org.jspecify.annotations.Nullable;

public interface UserRepository {

    Future<User> save(User user);

    Future<User> update(User user);

    Future<Optional<User>> findById(String id);

    Future<Optional<User>> findByUsername(String username);

    Future<List<User>> findAll();

    Future<List<User>> findByEmail(@Nullable String emailFilter);

    Future<Boolean> existsByUsername(String username);
}

File: apps/demo-be-ktkt/src/main/kotlin/com/organiclever/demoktkt/infrastructure/repositories/UserRepository.kt
package com.organiclever.demoktkt.infrastructure.repositories

import com.organiclever.demoktkt.domain.Page
import com.organiclever.demoktkt.domain.User
import com.organiclever.demoktkt.domain.UserStatus
import java.util.UUID

data class CreateUserRequest(
  val username: String,
  val email: String,
  val displayName: String,
  val passwordHash: String,
  val role: com.organiclever.demoktkt.domain.Role = com.organiclever.demoktkt.domain.Role.USER,
)

data class UpdateUserPatch(
  val displayName: String? = null,
  val passwordHash: String? = null,
  val status: UserStatus? = null,
  val failedLoginCount: Int? = null,
)

interface UserRepository {
  suspend fun findByUsername(username: String): User?

  suspend fun findById(id: UUID): User?

  suspend fun findByEmail(email: String): User?

  suspend fun create(request: CreateUserRequest): User

  suspend fun update(id: UUID, patch: UpdateUserPatch): User?

  suspend fun incrementFailedLogins(id: UUID): Int

  suspend fun resetFailedLogins(id: UUID)

  suspend fun findAll(page: Int, pageSize: Int, emailFilter: String?): Page<User>
}
